flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/CustomDistributionITCase.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.flink.util.Collector;
+import org.junit.Test;
+
+
+import java.io.IOException;
+
+import static org.junit.Assert.fail;
+
+
+public class CustomDistributionITCase {
+	
+	@Test
+	public void testPartitionWithDistribution1() throws Exception{
+		/*
+		 * Test the record partitioned rightly with one field according to the customized data distribution
+		 */
+
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input1 = CollectionDataSets.get3TupleDataSet(env);
+		final TestDataDist dist = new TestDataDist(1);
+
+		env.setParallelism(dist.getParallelism());
+
+		DataSet<Boolean> result = DataSetUtils.partitionByRange(input1, dist, 0).mapPartition(new RichMapPartitionFunction<Tuple3<Integer, Long, String>, Boolean>() {
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Long, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+				for (Tuple3<Integer, Long, String> s : values) {
+					if ((s.f0 - 1) / 7 != partitionIndex) {
+						fail("Record was not correctly partitioned: " + s.toString());
+					}
+				}
+			}
+		});
+
+		result.output(new DiscardingOutputFormat()); 
+		env.execute();
+	}
+
+	@Test
+	public void testRangeWithDistribution2() throws Exception{
+		/*
+		 * Test the record partitioned rightly with two fields according to the customized data distribution
+		 */
+
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input1 = CollectionDataSets.get3TupleDataSet(env);
+		final TestDataDist dist = new TestDataDist(2);
+
+		env.setParallelism(dist.getParallelism());
+
+		DataSet<Boolean> result = DataSetUtils.partitionByRange(input1.map(new MapFunction<Tuple3<Integer, Long, String>, Tuple3<Integer, Integer, String>>() {
+			@Override
+			public Tuple3<Integer, Integer, String> map(Tuple3<Integer, Long, String> value) throws Exception {
+				return new Tuple3<>(value.f0, value.f1.intValue(), value.f2);
+			}
+		}), dist, 0, 1).mapPartition(new RichMapPartitionFunction<Tuple3<Integer, Integer, String>, Boolean>() {
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Integer, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+				for (Tuple3<Integer, Integer, String> s : values) {
+					if (s.f0 <= partitionIndex * (partitionIndex + 1) / 2 ||
+							s.f0 > (partitionIndex + 1) * (partitionIndex + 2) / 2 ||
+							s.f1 - 1 != partitionIndex) {
+						fail("Record was not correctly partitioned: " + s.toString());
+					}
+				}
+			}
+		});
+
+		result.output(new DiscardingOutputFormat());
+		env.execute();
+	}
+
+	/**
+	 * The class is used to do the tests of range partition with customed data distribution.
+	 */
+	public static class TestDataDist implements DataDistribution {
```

Comment:
The two distributions defined by this class have not a lot in common. How about we define two classes, one for each distribution?